### ISSUE TITLE
General custom CRM development question

### ISSUE THREAD

**Issue by Tamas-Toth-ebola** [verygoodplugins/wp-fusion-lite#8]
Dear Developers!

Because our last some Mautic relevant PR requests maybe you already know something like that we are ( particularly with @pety-dc ) a small developer company who started to use your WP Fusion plugin as a paid version (with Profession plan) between our Mautic and WordPress instances. We love it and this is why @pety-dc made those some Mautic relevant bugfix PR-s what could ironing the available Mautic integration. For our WordPress and Mautic usage this seems perfect now and we will try to continuously watch this data connection because our WP <> Mautic configurations.

From another aspect we have a totally own e-mail marketing software/service in Hungary what is probably the largest and oldest in our country. Our e-mail marketing system ([MailEngine](https://www.mailengine.hu/)) - of course - is not a full featured CRM but has lot more features as a plain mail sender solution so based on the only one channel communication (e-mail) is has almost all features what need to a service for handle it's costumers/partners. Based on these informations in the last week in our head there was a spark that what if we would like to make an integration between WP Fusion and our MailEngine system.

@pety-dc saw that there are more possible way for it. There are already lot of CRMs in WP Fusion and because it maybe we could integrate our service into WP Fusion of course with your help. Or there is one custom CRM like integration possibility and with it maybe we could make one 'standalone' solution for our goal where the 'standalone' just suggest that in this case we do not (need to) try to code/extend the core plugin set but we could handle it with a WP Fusion 'external' plugin.

Currently our first question is that what is the preferred way for it? Could we start to make a core plugin what you could validate? Or we need to start to make one 'external' custom CRM integration based on your already available custom CRM integration solution?

If you could define the preferred way, our next question would be that what are the limits, rules, policies to make one such plugin? @pety-dc is currently trying to examine the available environment and codebase for the further steps.

Although our MailEngine service usually used only in our local country but would be very nice if with our development WP Fusion could expand it's CRL list with out service. Of course in this case all the responsibilities about the MailEngine integration could be ours as the e-mail marketing system is our and the WP Fusion integration also would be.

In MailEngine we have a SOAP API for external data connections so basically we would try o build the integration based on that SOAP interface but seems that in WP such restriction would be normal (as for example SOAP functions also need to be available for general WooCommerce usage).

Lot of thanks for your answer and effort in advance!

Tamas Toth (ebola)

**Comment by jack-arturo** [comment]
Hey @ebola-dc ,

Thanks! We really appreciate the PRs and having another set of eyes on our code :) 

I checked out the MailEngine site but I'm having a bit of a hard time reading about the API since it's not in English. But you can definitely create a custom CRM module.

We have a starter framework here which adds a new CRM to WP Fusion, and is contained in its own plugin: https://github.com/verygoodplugins/wp-fusion-custom-crm

We have some info on creating custom CRM modules here: https://wpfusion.com/documentation/advanced-developer-tutorials/creating-custom-crm-modules/

You'd basically just go through all the methods in the CRM class at https://github.com/verygoodplugins/wp-fusion-custom-crm/blob/master/includes/class-wpf-custom.php and update them to use the proper authentication and API calls for MailEngine, and to return the proper responses. You can see examples in our other CRM classes, and we also have a document explaining the core methods here https://wpfusion.com/documentation/advanced-developer-tutorials/how-wp-fusion-interfaces-with-multiple-crms/

There aren't really any limits / rules regarding extending WP Fusion. We're proudly open source / GPL and are happy to see people extend it.

If you're able to complete the integration and you're willing to share it with us, we could consider including it in WP Fusion core. Or if you'd prefer we could keep the module for MailEngine separate, for example if you wanted to sell it yourself. That'd be fine too.

Good luck! If you need any help or want someone to look over your code you can also reach us at support@wpfusionplugin.com.

**Comment by Tamas-Toth-ebola** [comment]
@verygoodplugins thanks for your detailed answer. First of all sorry about the Hungarian SOAP API documentation, at the beginning we had it in 2 language (HUN and ENG) but as later there were not too much clients who had need the English one we simply did not continue our version tracking/translation of the HUN reference doc.

About the technical details @pety-dc could be consult with you (us) in case if he needed.

Until yet, thanks for your help and answer.

**Comment by pety-dc** [comment]
Hey @verygoodplugins ,

We have finished creating a Mailengine integration CRM extension for WP fusion.
The github repo have English readme, you may [check it here](https://github.com/d-code-ltd/wp-fusion-extension-mailengine-crm) 

Please, consider to add this CRM to the WP Fusion core. Could you tell me what validation steps should we pass in order to achvieve that?

**Comment by jack-arturo** [comment]
Hey @pety-dc, @ebola-dc 

Sorry for the delay. We got swamped with Black Friday stuff and then the holidays came, I didn't get a proper chance to look at this.

Looks great! Really nice work. I wasn't able to test it fully without a MailEngine account but the code looks good and I can see what you're doing. I've included that CRM module into WPF core for v3.28, which will be released today.

I'm pretty sure you won't have any conflicts between your custom plugin and WPF core, since the class is only supposed to be instantiated once, but just be careful when you install v3.28 to make sure nothing clashes.

Thanks again, and happy new year! :)
